nextstrain/cli/remote/nextstrain_dot_org.py

```diff
@@ -0,0 +1,656 @@
+"""
+Nextstrain.org remote.
+
+Backend module for the remote family of commands.
+
+.. envvar:: NEXTSTRAIN_DOT_ORG
+
+    Base URL to use instead of ``https://nextstrain.org`` when accessing
+    remote resources.  Useful primarily for development and testing, e.g.:
+
+    .. code-block:: shell
+
+        NEXTSTRAIN_DOT_ORG=http://localhost:5000 nextstrain remote ls nextstrain.org
+"""
+
+import os
+import requests
+import urllib.parse
+from collections import defaultdict
+from email.message import EmailMessage
+from pathlib import Path, PurePosixPath
+from textwrap import indent
+from typing import Dict, Iterable, List, NamedTuple, NewType, Optional, Tuple, Union
+from urllib.parse import urljoin, urlsplit, quote as urlquote, quote_plus as urlquote_plus
+from ..authn import current_user
+from ..errors import UserError
+from ..gzip import GzipCompressingReader
+from ..util import remove_prefix, glob_matcher, glob_match
+
+
+NEXTSTRAIN_DOT_ORG = os.environ.get("NEXTSTRAIN_DOT_ORG") \
+                  or "https://nextstrain.org"
+
+
+# This subtype lets us use the type checker to catch places where code
+# expects/assumes the semantics of normalize_path().
+class NormalizedPath(PurePosixPath):
+    ...
+
+
+class Resource:
+    """
+    Base class for a remote Nextstrain resource, as described by a Charon API
+    "getAvailable" response.
+
+    Concretely, either a :class:`Dataset` or :class:`Narrative` currently.
+    """
+    path: NormalizedPath
+    subresources: List['SubResource']
+
+    def __init__(self, api_item: dict):
+        self.path = normalize_path(api_item["request"])
+
+
+class SubResource(NamedTuple):
+    """
+    Info about a dataset or narrative subresource (i.e. file), representing
+    both primary data and sidecars.
+
+    media_type : str
+        nextstrain.org RESTful API media type.
+
+    file_extension : str
+        File extension to use when saving this subresource to a file.
+
+    primary : bool
+        Flag indicating if this subresource is the resource's primary data.
+    """
+    media_type: str
+    file_extension: str
+    primary: bool = False
+
+
+class Dataset(Resource):
+    """
+    A remote Nextstrain dataset, as described by a Charon API response,
+    extended for the nextstrain.org RESTful API.
+    """
+    def __init__(self, api_item):
+        super().__init__(api_item)
+
+        default_sidecars = ["root-sequence", "tip-frequencies"]
+
+        self.subresources = [
+            SubResource("application/vnd.nextstrain.dataset.main+json", ".json", primary = True),
```

Comment:
Maybe? My thinking here was that media types are the kind of constants where the value _is_ the name and vice versa, like, say, HTTP methods. That is, they're different from constants where the value is not synonymous with the name.

For this type of self-describing constant, I think they're more clearly used directly rather than indirectly behind a translation layer (dict/enum) that gives them some other name. My thinking goes, why have multiple names, varying across code bases, for a thing when you can have one that's consistent across codebases?

One benefit of indirection with an enum (but not dict) is that it would let the type checker tell us if we made a typo, but that feels like a marginal gain not worth the indirection to me. (The code will also likely fail to work with a typo in the media type, so would be noticeable.)